subsys/mgmt/osdp/src/osdp_pd.c

```diff
@@ -174,125 +247,132 @@ static int pd_decode_command(struct osdp_pd *pd, uint8_t *buf, int len)
+		cmd.buzzer.control_code = buf[pos++];
+		cmd.buzzer.on_count = buf[pos++];
+		cmd.buzzer.off_count = buf[pos++];
+		cmd.buzzer.rep_count = buf[pos++];
+		ret = pd->command_callback(pd->command_callback_arg, &cmd);
+		if (ret != 0) {
+			pd->reply_id = REPLY_NAK;
+			pd->cmd_data[0] = OSDP_PD_NAK_RECORD;
+			ret = OSDP_PD_ERR_REPLY;
 			break;
 		}
-		cmd->id = OSDP_CMD_BUZZER;
-		cmd->buzzer.reader       = buf[pos++];
-		cmd->buzzer.control_code = buf[pos++];
-		cmd->buzzer.on_count     = buf[pos++];
-		cmd->buzzer.off_count    = buf[pos++];
-		cmd->buzzer.rep_count    = buf[pos++];
-		osdp_cmd_enqueue(pd, cmd);
 		pd->reply_id = REPLY_ACK;
 		ret = OSDP_PD_ERR_NONE;
 		break;
 	case CMD_TEXT:
-		if (len < CMD_TEXT_DATA_LEN) {
+		if (len < CMD_TEXT_DATA_LEN || !pd->command_callback) {
 			break;
 		}
-		cmd = osdp_cmd_alloc(pd);
-		if (cmd == NULL) {
-			LOG_ERR("cmd alloc error");
+		cmd.id = OSDP_CMD_TEXT;
+		cmd.text.reader = buf[pos++];
+		cmd.text.control_code = buf[pos++];
+		cmd.text.temp_time = buf[pos++];
+		cmd.text.offset_row = buf[pos++];
+		cmd.text.offset_col = buf[pos++];
+		cmd.text.length = buf[pos++];
+		if (cmd.text.length > OSDP_CMD_TEXT_MAX_LEN ||
+		    ((len - CMD_TEXT_DATA_LEN) < cmd.text.length) ||
+		    cmd.text.length > OSDP_CMD_TEXT_MAX_LEN) {
 			break;
 		}
-		cmd->id = OSDP_CMD_TEXT;
-		cmd->text.reader       = buf[pos++];
-		cmd->text.control_code = buf[pos++];
-		cmd->text.temp_time    = buf[pos++];
-		cmd->text.offset_row   = buf[pos++];
-		cmd->text.offset_col   = buf[pos++];
-		cmd->text.length       = buf[pos++];
-		if (cmd->text.length > OSDP_CMD_TEXT_MAX_LEN ||
-		    ((len - CMD_TEXT_DATA_LEN) < cmd->text.length) ||
-		    cmd->text.length > OSDP_CMD_TEXT_MAX_LEN) {
-			osdp_cmd_free(pd, cmd);
-			break;
+		for (i = 0; i < cmd.text.length; i++) {
+			cmd.text.data[i] = buf[pos++];
 		}
-		for (i = 0; i < cmd->text.length; i++) {
-			cmd->text.data[i] = buf[pos++];
+		ret = pd->command_callback(pd->command_callback_arg, &cmd);
+		if (ret != 0) {
+			pd->reply_id = REPLY_NAK;
+			pd->cmd_data[0] = OSDP_PD_NAK_RECORD;
+			ret = OSDP_PD_ERR_REPLY;
+			break;
 		}
-		osdp_cmd_enqueue(pd, cmd);
 		pd->reply_id = REPLY_ACK;
 		ret = OSDP_PD_ERR_NONE;
 		break;
 	case CMD_COMSET:
-		if (len != CMD_COMSET_DATA_LEN) {
-			break;
+		if (len != CMD_COMSET_DATA_LEN || !pd->command_callback) {
+			break;
+		}
+		cmd.id = OSDP_CMD_COMSET;
+		cmd.comset.address = buf[pos++];
+		cmd.comset.baud_rate = buf[pos++];
+		cmd.comset.baud_rate |= buf[pos++] << 8;
+		cmd.comset.baud_rate |= buf[pos++] << 16;
+		cmd.comset.baud_rate |= buf[pos++] << 24;
+		if (cmd.comset.address >= 0x7F ||
+		    (cmd.comset.baud_rate != 9600 &&
+		     cmd.comset.baud_rate != 19200 &&
+		     cmd.comset.baud_rate != 38400 &&
+		     cmd.comset.baud_rate != 115200 &&
+		     cmd.comset.baud_rate != 230400)) {
+			LOG_ERR("COMSET Failed! command discarded");
+			cmd.comset.address = pd->address;
+			cmd.comset.baud_rate = pd->baud_rate;
```

Comment:
Should be called pd->command_callback in case when command is discarded?

Reply:
Good catch, we can avoid the callback when the command is discarded.